### PR TITLE
Coverity issues involving negative arguments

### PR DESCRIPTION
**Pull request by paulidale**
- [ ] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by t8m** [comment]
Was this closed intentionally? Although maybe we should just change those _size() calls to return 0 instead of negative values.

**Comment by paulidale** [comment]
Not, I must have closed it by accident.  Lots of these outstanding at present.

**Comment by t8m** [comment]
I really wonder if introducing the negative return value for these calls was right.

**Comment by paulidale** [comment]
These are mostly historic calls, so we have to live with them.
A new call that did something like this would be different.  Then again, a new call that returns a size as an int is suspect from the start.

File: crypto/dh/dh_pmeth.c [hunk 1]
```diff
@@ -463,11 +463,10 @@ static int pkey_dh_derive(EVP_PKEY_CTX *ctx, unsigned char *key,
...
         ret = 0;
-        Zlen = DH_size(dh);
-        Z = OPENSSL_malloc(Zlen);
-        if (Z == NULL) {
-            goto err;
-        }
+        if ((Zlen = DH_size(dh)) <= 0)
+            return 0;
+        if ((Z = OPENSSL_malloc(Zlen)) == NULL)
+            return 0;
```
Comment (FdaSilvaYY) [inline crypto/dh/dh_pmeth.c:469]: Suggestion: may be add a call to `ERR_raise(ERR_LIB_BN, ERR_R_MALLOC_FAILURE);` here.

Comment (paulidale) [reply crypto/dh/dh_pmeth.c:470]: DWIW: I've added the error message.  There are other cases in the code where errors aren't fully reported and where errors aren't possible but are checked for.

**Comment by paulidale** [comment]
ping for review

**Comment by paulidale** [comment]
Still feeling unrequited review.

**Comment by t8m** [comment]
I'm still a little bit worried for the cases where we are newly introducing the negative error return value in 3.0. This can have quite serious impact in applications if they aren't properly ready to handle negative error returns.

**Comment by paulidale** [comment]
This isn't introducing any such functions.  I don't think it's checking the returns from any either.

`DH_size()`, `EVP_CIPHER_CTX_key_length()` and `EVP_CIPHER_CTX_iv_length()` are all pre-existing.


I agree, it's a concern and we really shouldn't be doing it.

**Comment by t8m** [comment]
> `DH_size()`, `EVP_CIPHER_CTX_key_length()` and `EVP_CIPHER_CTX_iv_length()` are all pre-existing.

I do not think `DH_size()` and `EVP_CIPHER_CTX_key_length()` can ever return < 0 values in 1.1.1.
`EVP_CIPHER_CTX_iv_length()` is pre-existing.

I am not saying _this_ PR is introducing it. I am saying 3.0 introduces it. So maybe the right fix for this PR is to actually drop the negative returns for these.

**Comment by paulidale** [comment]
In essence, you are suggesting returning zero from `EVP_CIPHER_CTX_key_length()` instead of the `EVP_CTRL_RET_UNSUPPORTED` value.  I'm not convinced that this is a good idea.  0 or negative returns are both going to cause problems.

All we're saving are the programs that check `== 0` instead of `<= 0`.

Our documentation isn't forgiving here: negative and zero returns are not mentioned.

**Comment by FdaSilvaYY** [comment]
Ping @openssl ; This PR is falling into the limbo.

**Comment by paulidale** [comment]
Rebased to latest master, still waiting for a review or a suggestion about the negative return changes required.

**Review by mattcaswell — APPROVED** [review]
This looks fine to me. This PR isn't introducing any changes which causes functions to return <0 where they didn't before. So, if that has occurred it is out of scope of this PR to fix it IMO.

**Comment by paulidale** [comment]
Merged to master, thanks for the comments etc.

**Pull request closed without merging**